Code/GraphMol/Wrap/Sgroup.cpp

```diff
@@ -0,0 +1,102 @@
+
+std::string sGroupClassDoc =
+    "A collection of atoms and bonds with associated properties\n";
+
+struct sgroup_wrap {
+  static void wrap() {
+    python::class_<SGroup, boost::shared_ptr<SGroup>>(
+        "SGroup", sGroupClassDoc.c_str(), python::no_init)
+        .def("GetOwningMol", &SGroup::getOwningMol,
+             "returns the molecule owning this SGroup",
+             python::return_internal_reference<>())
+        .def("GetIndexInMol", &SGroup::getIndexInMol,
+             "returns the index of this SGroup in the owning molecule's list.")
+        .def("GetAtoms", &SGroup::getAtoms,
+             "returns a list of the indices of the atoms in this SGroup",
+             python::return_value_policy<python::copy_const_reference>())
+        .def("GetParentAtoms", &SGroup::getParentAtoms,
+             "returns a list of the indices of the parent atoms in this SGroup",
+             python::return_value_policy<python::copy_const_reference>())
+        .def("GetBonds", &SGroup::getBonds,
+             "returns a list of the indices of the bonds in this SGroup",
+             python::return_value_policy<python::copy_const_reference>())
+        .def("HasProp",
+             (bool (RDProps::*)(const std::string &) const) & SGroup::hasProp,
+             "returns whether or not a particular property exists")
+        .def("GetProp",
+             (std::string(RDProps::*)(const std::string &) const) &
+                 SGroup::getProp<std::string>,
+             "returns the value of a particular property")
+        .def("GetIntProp",
+             (int (RDProps::*)(const std::string &) const) &
+                 SGroup::getProp<int>,
+             "returns the value of a particular property")
+        .def("GetUnsignedProp",
+             (unsigned int (RDProps::*)(const std::string &) const) &
+                 SGroup::getProp<unsigned int>,
+             "returns the value of a particular property")
+        .def("GetDoubleProp",
+             (double (RDProps::*)(const std::string &) const) &
+                 SGroup::getProp<double>,
+             "returns the value of a particular property")
+        .def("GetBoolProp",
+             (bool (RDProps::*)(const std::string &) const) &
+                 SGroup::getProp<bool>,
+             "returns the value of a particular property")
+        .def("GetPropNames", &SGroup::getPropList,
+             (python::arg("self"), python::arg("includePrivate") = false,
+              python::arg("includeComputed") = false),
+             "Returns a list of the properties set on the SGroup.\n\n")
+        .def("GetPropsAsDict", GetPropsAsDict<SGroup>,
+             (python::arg("self"), python::arg("includePrivate") = true,
+              python::arg("includeComputed") = true),
+             "Returns a dictionary of the properties set on the SGroup.\n"
+             " n.b. some properties cannot be converted to python types.\n");
+    python::def("GetMolSGroups", &getMolSGroups,

```

Comment:
I think this should use:

```
return_internal_reference
```

so that the owning molecule doesn't get deleted while someone is holding onto the sgroups.

Reply:
It's not an internal reference. We create a new python object that contains the pointers. I don't think it's possible to put lifetime restrictions on those pointers using boost.
There are a couple of spots where this is a problem in the code and I don't think there's a great solution to it.

Reply:
We could do something like what Mol.GetAtoms() does:
https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/Wrap/Mol.cpp#L690
and
https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/Wrap/Mol.cpp#L127

I think that's effective, but it's pretty horrible since it produces these things:
``` 
In [3]: ats = m.GetAtoms()                                                                                                        

In [4]: ats                                                                                                                       
Out[4]: <rdkit.Chem.rdchem._ROAtomSeq at 0x7f661402eb20>
```

Reply:
I will play around a little bit and see if I can figure something out

Reply:
Solved this using a vector_indexing_suite. This should allow us to make the SGroups modifiable from Python, but we can do that in a later PR (or future release).